File: cpp/tests/codegen/test_timestamp.cpp
/*
 * Copyright (C) 2019 Geon Technologies, LLC
 *
 * This file is part of vrtgen.
 *
 * vrtgen is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * vrtgen is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 */

#include "catch/catch.hpp"

#include "timestamp.hpp"

#include <vector>

TEST_CASE("UTC picosecond-resolution timestamp")
{
    TimestampUTCRealTime packet_in;
    packet_in.setIntegerTimestamp(10000000);
    packet_in.setFractionalTimestamp(999999999999);

    size_t packed_size = packing::TimestampUTCRealTimeHelper::bytes_required(packet_in);
    CHECK(packed_size == 24);

    std::vector<unsigned char> data;
    data.resize(24);
    packing::TimestampUTCRealTimeHelper::pack(packet_in, data.data(), data.size());

    const vrtgen::packing::Header* header = reinterpret_cast<const vrtgen::packing::Header*>(data.data());
    CHECK(header->getTSI() == vrtgen::TSI::UTC);
    CHECK(header->getTSF() == vrtgen::TSF::REAL_TIME);

    CHECK(packing::TimestampUTCRealTimeHelper::match(data.data(), data.size()));

    TimestampUTCRealTime packet_out;
    packing::TimestampUTCRealTimeHelper::unpack(packet_out, data.data(), data.size());
    CHECK(packet_out.getIntegerTimestamp() == 10000000);
    CHECK(packet_out.getFractionalTimestamp() == 999999999999);
}

TEST_CASE("GPS seconds timestamp")
{
    TimestampGPS packet_in;
    packet_in.setIntegerTimestamp(12345678);
 
    size_t packed_size = packing::TimestampGPSHelper::bytes_required(packet_in);
    CHECK(packed_size == 16);

    std::vector<unsigned char> data;
    data.resize(16);
    packing::TimestampGPSHelper::pack(packet_in, data.data(), data.size());

    const vrtgen::packing::Header* header = reinterpret_cast<const vrtgen::packing::Header*>(data.data());
    CHECK(header->getTSI() == vrtgen::TSI::GPS);
    CHECK(header->getTSF() == vrtgen::TSF::NONE);

    CHECK(packing::TimestampGPSHelper::match(data.data(), data.size()));

    TimestampGPS packet_out;
    packing::TimestampGPSHelper::unpack(packet_out, data.data(), data.size());
    CHECK(packet_out.getIntegerTimestamp() == 12345678);
}

TEST_CASE("Free-running fractional timestamp")
{
    TimestampFreeRunning packet_in;
    packet_in.setFractionalTimestamp(0x1E1E1E1E1E1E1E1E);

    size_t packed_size = packing::TimestampFreeRunningHelper::bytes_required(packet_in);
    CHECK(packed_size == 20);

    std::vector<unsigned char> data;
    data.resize(20);
    packing::TimestampFreeRunningHelper::pack(packet_in, data.data(), data.size());

    const vrtgen::packing::Header* header = reinterpret_cast<const vrtgen::packing::Header*>(data.data());
    CHECK(header->getTSI() == vrtgen::TSI::NONE);
    CHECK(header->getTSF() == vrtgen::TSF::FREE_RUNNING);

    CHECK(packing::TimestampFreeRunningHelper::match(data.data(), data.size()));

    TimestampFreeRunning packet_out;
    packing::TimestampFreeRunningHelper::unpack(packet_out, data.data(), data.size());
    CHECK(packet_out.getFractionalTimestamp() == 0x1E1E1E1E1E1E1E1E);
}

File: tests/parser/test_timestamp.py
# Copyright (C) 2022 Geon Technologies, LLC
#
# This file is part of vrtgen.
#
# vrtgen is free software: you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option)
# any later version.
#
# vrtgen is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
# more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.

import yaml
import pytest
from helpers.utils import parse_document
from vrtgen.parser.model.types import enums
from vrtgen.parser.loader import get_loader
from vrtgen.parser.model.types.enums import *

def check_tsi(timestamp, value):
    if value == 'none':
        assert timestamp.tsi.value == 0
    elif value == 'utc':
        assert timestamp.tsi.value == 1
    elif value == 'gps':
        assert timestamp.tsi.value == 2
    elif value == 'other':
        assert timestamp.tsi.value == 3

def check_tsf(timestamp, value):
    if value == 'none':
        assert timestamp.tsf.value == 0
    elif value == 'sample_count':
        assert timestamp.tsf.value == 1
    elif value == 'picoseconds':
        assert timestamp.tsf.value == 2
    elif value == 'free-running':
        assert timestamp.tsf.value == 3

def check_basics(name, timestamp):
    assert name == 'timestamp'
    assert timestamp.name == 'timestamp'
    assert timestamp.type_ == 'Timestamp'
    assert timestamp.enabled
    assert timestamp.required

def test_timestamp_default():
    document = """
    timestamp: !Timestamp
    """
    name,timestamp = parse_document(document)
    check_basics(name, timestamp)
    assert timestamp.bits == 0
    # Integer
    assert not timestamp.integer.enabled
    assert not timestamp.integer.required
    assert not timestamp.integer.value
    # TSI
    assert not timestamp.tsi.enabled
    assert not timestamp.tsi.required
    assert not timestamp.tsi.value
    # Fractional
    assert not timestamp.fractional.enabled
    assert not timestamp.fractional.required
    assert not timestamp.fractional.value
    # TSF
    assert not timestamp.tsf.enabled
    assert not timestamp.tsf.required
    assert not timestamp.tsf.value

testdata = [
    ('integer', 'none'),
    ('integer', 'utc'),
    ('integer', 'gps'),
    ('integer', 'other')
]

@pytest.mark.parametrize('parameter,value', testdata)
def test_timestamp_integer(parameter, value):
    document = """
    timestamp: !Timestamp
        {}: {}
    """.format(parameter, value)
    name,timestamp = parse_document(document)
    check_basics(name, timestamp)
    assert timestamp.bits == 32 # Integer part with no Fractional
    # Integer
    assert timestamp.integer.enabled
    assert timestamp.integer.required
    assert not timestamp.integer.value
    # TSI
    assert not timestamp.tsi.enabled
    assert not timestamp.tsi.required
    check_tsi(timestamp, value)
    # Fractional
    assert not timestamp.fractional.enabled
    assert not timestamp.fractional.required
    assert not timestamp.fractional.value
    # TSF
    assert not timestamp.tsf.enabled
    assert not timestamp.tsf.required
    assert timestamp.tsf.value == enums.TSF.NONE

testdata = [
    ('fractional', 'none'),
    ('fractional', 'samples'),
    ('fractional', 'picoseconds'),
    ('fractional', 'free-running')
]

@pytest.mark.parametrize('parameter,value', testdata)
def test_timestamp_fractional(parameter, value):
    document = """
    timestamp: !Timestamp
        {}: {}
    """.format(parameter, value)
    name,timestamp = parse_document(document)
    check_basics(name, timestamp)
    assert timestamp.bits == 64 # Fractional part with no Integer
    # Integer
    assert not timestamp.integer.enabled
    assert not timestamp.integer.required
    assert not timestamp.integer.value
    # TSI
    assert not timestamp.tsi.enabled
    assert not timestamp.tsi.required
    assert timestamp.tsi.value == 0
    # Fractional
    assert timestamp.fractional.enabled
    assert timestamp.fractional.required
    assert not timestamp.fractional.value
    # TSF
    assert not timestamp.tsf.enabled
    assert not timestamp.tsf.required
    check_tsf(timestamp, value)

def test_timestamp_both():
    integer_value = 'utc'
    fractional_value = 'picoseconds'
    document = """
    timestamp: !Timestamp
        integer: {}
        fractional: {}
    """.format(integer_value, fractional_value)
    name,timestamp = parse_document(document)
    check_basics(name, timestamp)
    assert timestamp.bits == 32 + 64 # Integer and Fractional
    # Integer
    assert timestamp.integer.enabled
    assert timestamp.integer.required
    assert not timestamp.integer.value
    # TSI
    assert not timestamp.tsi.enabled
    assert not timestamp.tsi.required
    check_tsi(timestamp, integer_value)
    # Fractional
    assert timestamp.fractional.enabled
    assert timestamp.fractional.required
    assert not timestamp.fractional.value
    # TSF
    assert not timestamp.tsf.enabled
    assert not timestamp.tsf.required
    check_tsf(timestamp, fractional_value)

def test_timestamp_invalid_key():
    document = """
    timestamp: !Timestamp
        invalid_key: test
    """
    with pytest.raises(KeyError) as e_info:
        parse_document(document)

def test_timestamp_invalid_integer():
    document = """
    timestamp: !Timestamp
        integer: invalid
    """
    with pytest.raises(ValueError) as e_info:
        parse_document(document)

def test_timestamp_invalid_fractional():
    document = """
    timestamp: !Timestamp
        fractional: invalid
    """
    with pytest.raises(ValueError) as e_info:
        parse_document(document)